packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,201 @@
+error RelayMessageCallerNotCrossL2Inbox();
+
+/// @notice Thrown when attempting to relay a message where CrossL2Inbox's origin is not L2ToL2CrossDomainMessenger.
+error CrossL2InboxOriginNotL2ToL2CrossDomainMessenger();
+
+/// @notice Thrown when attempting to relay a message whose destination chain is not the chain relaying it.
+error MessageDestinationNotRelayChain();
+
+/// @notice Thrown when attempting to relay a message whose target is CrossL2Inbox.
+error MessageTargetCrossL2Inbox();
+
+/// @notice Thrown when attempting to relay a message whose target is L2ToL2CrossDomainMessenger.
+error MessageTargetL2ToL2CrossDomainMessenger();
+
+/// @notice Thrown when attempting to relay a message that has already been relayed.
+error MessageAlreadyRelayed();
+
+/// @custom:proxied
+/// @custom:predeploy 0x4200000000000000000000000000000000000023
+/// @title L2ToL2CrossDomainMessenger
+/// @notice The L2ToL2CrossDomainMessenger is a higher level abstraction on top of the CrossL2Inbox that provides
+///         features necessary for secure transfers ERC20 tokens between L2 chains. Messages sent through the
+///         L2ToL2CrossDomainMessenger on the source chain receive both replay protection as well as domain binding.
+contract L2ToL2CrossDomainMessenger is IL2ToL2CrossDomainMessenger, ISemver, TransientReentrancyAware {
+    /// @notice Storage slot for the sender of the current cross domain message.
+    ///         Equal to bytes32(uint256(keccak256("l2tol2crossdomainmessenger.sender")) - 1)
+    bytes32 internal constant CROSS_DOMAIN_MESSAGE_SENDER_SLOT =
+        0xb83444d07072b122e2e72a669ce32857d892345c19856f4e7142d06a167ab3f3;
+
+    /// @notice Storage slot for the source of the current cross domain message.
+    ///         Equal to bytes32(uint256(keccak256("l2tol2crossdomainmessenger.source")) - 1)
+    bytes32 internal constant CROSS_DOMAIN_MESSAGE_SOURCE_SLOT =
+        0x711dfa3259c842fffc17d6e1f1e0fc5927756133a2345ca56b4cb8178589fee7;
+
+    /// @notice Current message version identifier.
+    uint16 public constant MESSAGE_VERSION = uint16(0);
+
+    /// @notice Semantic version.
+    /// @custom:semver 1.0.0
+    string public constant version = "1.0.0";
+
+    /// @notice Mapping of message hashes to boolean receipt values. Note that a message will only be present in this
+    ///         mapping if it has successfully been relayed on this chain, and can therefore not be relayed again.
+    mapping(bytes32 => bool) public successfulMessages;
+
+    /// @notice Nonce for the next message to be sent, without the message version applied. Use the messageNonce getter,
+    ///         which will insert the message version into the nonce to give you the actual nonce to be used for the
+    ///         message.
+    uint240 internal msgNonce;
+
+    /// @notice Emitted whenever a message is sent to the other chain.
+    /// @param data Encoded data of the message that was sent.
+    event SentMessage(bytes data);
```

Comment:
this needs to be explicitly marked as `anonymous` so there are 0 topics, otherwise the topic will clash with the function signatures. We would need to use a custom 32 byte jump table based on the first topic if we wanted to encode cross domain calls this way

Reply:
unfortunately, foundry doesn't recognize anonymous events without an `indexer` field: https://github.com/foundry-rs/foundry/issues/7457. will have to add indexer to make it work!

Reply:
Perhaps we can use the `vm.recordLogs` and `vm.stopRecordingAndGetLogs` (not sure of exact name) cheatcodes to handle this and then assert directly on the serialized log. This is actually a better test because the log should be serialized in consensus format (which then plays into the interop msg serialization) and we will have test coverage of the log0 scheme being sound

Reply:
good idea!